mantidproject/mantid
# Add algorithm to sum banks for ISIS Reflectometry

**Description of work.**

This PR adds an algorithm to do the first steps of reduction from an ISIS Reflectometry 2D detector to sum across banks from a region of interest into a single bank.

This uses SmoothNeighbours to do the summing. Some minor changes were required to get this to work with INTER data:
  - Ignore slits and supermirrors in `containsRectDetectors` because they were causing the instrument to be incorrectly identified as non-rectangular.
  - Add a flag to `getDetectorIDToWorkspaceIndexMap` to ignore invalid detector IDs. This was causing an error due to having multiple detectors per spectrum, but this is irrelevant in our case where those are invalid detectors which will be ignored anyway.

**To test:**

- Download the input workspace here (note that this has had an experimental IDF applied to it).
  - [INTER45455_inst.zip](https://github.com/mantidproject/mantid/files/7453399/INTER45455_inst.zip)
- Run the script below. Note that the first two lines execute the new algorithm. The rest is a comparison with the same calculation done using GroupDetectors.
- Check that the CompareWorkspaces call at the end outputs the text `The workspaces "INTER45455_summedX" and "Inter45455_SummedX_GroupDets" matched!` in the log.

```
ws=Load('INTER45455_inst.nxs', OutputWorkspace='INTER45455')
ReflectometryISISSumBanks(InputWorkspace='INTER45455', OutputWorkspace='INTER45455_summedX', ROIDetectorIDs='2061-2180')

#
# Manual check against sum in X using GroupDetectors
#
start_idx, end_idx = 65, 184 # ws idx for our start/end det ID
npixels=ws.getInstrument().getComponentByName('linear-detector').ypixels()
groups = [""] * npixels
for ws_idx in range(ws.getNumberHistograms()):
    if ws_idx < start_idx or ws_idx > end_idx:
        continue
    # Work out which output pixel to sum the input pixel into
    bank = math.ceil(ws_idx / npixels)
    pixel = ws_idx - bank * npixels
    # Create a grouping pattern to do the sum for each output pixel
    groups[pixel] = groups[pixel] + str(ws_idx) + '+'
# Create the final grouping pattern to do all the sums
grouping = ', '.join(groups)
GroupDetectors(InputWorkspace=ws, OutputWorkspace='Inter45455_SummedX_GroupDets', GroupingPattern=grouping, Behaviour='Sum')
CompareWorkspaces(Workspace1='INTER45455_summedX', Workspace2='Inter45455_SummedX_GroupDets', CheckSpectraMap=False, CheckInstrument=False, CheckSample=True, ToleranceRelErr=True, CheckAllData=True)
```

This sums the middle two banks pixel-by-pixel into a single bank. Sliceviewer plots of the original and summed workspaces should look as follows:

Before:

![image](https://user-images.githubusercontent.com/12895056/139676374-2e010dfb-770e-4685-a255-0a49b3c47348.png)

After:

![image](https://user-images.githubusercontent.com/12895056/139676420-e301e3d2-f55a-4f59-befb-9436e3484b01.png)

Fixes #32771

Release notes to follow


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
